src/array_api_typing/_array.py

```diff
@@ -0,0 +1,88 @@
+__all__ = (
+    "Array",
+    "BoolArray",
+    "HasArrayNamespace",
+    "NumericArray",
+)
+
+from pathlib import Path
+from types import ModuleType
+from typing import Literal, Never, Protocol, TypeAlias
+from typing_extensions import TypeVar
+
+import optype as op
+
+from ._utils import docstring_setter
+
+# Load docstrings from TOML file
+try:
+    import tomllib
+except ImportError:
+    import tomli as tomllib  # type: ignore[import-not-found, no-redef]
+
+_docstrings_path = Path(__file__).parent / "_array_docstrings.toml"
+with _docstrings_path.open("rb") as f:
+    _array_docstrings = tomllib.load(f)["docstrings"]
+
+NS_co = TypeVar("NS_co", covariant=True, default=ModuleType)
+T_contra = TypeVar("T_contra", contravariant=True)
+R_co = TypeVar("R_co", covariant=True, default=Never)
+
+
+class HasArrayNamespace(Protocol[NS_co]):
+    """Protocol for classes that have an `__array_namespace__` method.
```

Comment:
Maybe we should mention that this is only intended for static typing, not for runtime things. And maybe also that this is intended as purely structural type, not a nominal one (i.e. not intended as base class).